### PR TITLE
Always run OptiType on tumor/normal DNA sequencing data

### PR DESCRIPTION
**Pull request by armish**
Seq2HLA runs are optional and depend on the RNA data availability, but running OptiType on DNA sequencing data might come handy for non-RNA-centered neoepitope predictions.

### CODE REVIEW

File: src/lib/pipeline.ml [hunk 1]
```diff
@@ -302,6 +306,8 @@ module Full (Bfx: Extended_edsl.Semantics) = struct
       ) in
     let normal = Stdlib.fastq_of_input parameters.normal in
     let tumor = Stdlib.fastq_of_input parameters.tumor in
+    let normal_tumor = Bfx.list [normal; tumor;] |> Bfx.concat in
+    let optitype = optitype_dna normal_tumor in
```
Comment (ihodes) [inline src/lib/pipeline.ml]: should it be running on both normal and tumor DNA? seems reason to run on normal, or both, but confusing to run on both

File: src/lib/pipeline.ml [hunk 2]
```diff
@@ -74,6 +74,7 @@ module Parameters = struct
     with_mutect2: bool [@default false];
     with_varscan: bool [@default false];
     with_somaticsniper: bool [@default false];
+    with_optitype: bool [@default false];
```
Comment (ihodes) [inline src/lib/pipeline.ml]: I like that it's optional here, but you need to implement that logic below (and through the report, as with `rna_bam` etc. Fine if it runs by default, though `[@default true]`. Not sure OTOMH how to resolve the confusion between seq2hla's results being fed through the rest of the pipeline, but optitype's being completely ignored (except for the sake of the report). 

One idea (and 100% fine with it not being addressed here, but if not, could you file an issue?), is to add a CLI option like `--hla-type-for-epitope-prediction` as an enum type one of [Optitype, Seq2Hla, Explicit (i.e. we pass the HLA types to the pipeline directly)] with some sort of fail-fast runtime error check that those options exist (i.e. if we specify optitype, that we're running optitype, likewise for Seq2Hla, and similarly that if we specify "explicit", that we're passing HLA types on the CLI). There may be a cleaner way to handle this, though; adding more and more CLI flags begins to get scary, quickly…

**Review by ihodes — CHANGES_REQUESTED** [review]
Nice addition, couple comments!

File: src/lib/pipeline.ml (hunk 2 above)
Comment (hammer) [reply src/lib/pipeline.ml]: > Not sure OTOMH how to resolve the confusion between seq2hla's results being fed through the rest of the pipeline

We could have some logic to detect if multiple HLA typers are run, compare the results, and highlight when we have a mismatch?

Of course we'll need logic on what to do upon mismatch--we could have a ranking of HLA typers by those we trust most I guess. I actually think OptiType is more accurate than seq2HLA from what I recall...

Comment (ihodes) [reply src/lib/pipeline.ml]: > We could have some logic to detect if multiple HLA typers are run, compare the results, and highlight when we have a mismatch?

This is a nice longer-term solution, and could be generally useful itself as well as we add more typers.

Comment (rleonid) [reply src/lib/pipeline.ml]: To throw another monkey wrench into this but OptiType can run on both RNA and DNA

**Comment by armish** [comment]
All right - I refactored things a bit to be able to support running optitype on normal, tumor and/or RNA samples. I haven't parametrize which HLA option is going to be used for the neoepitope prediction, but I would like address that with a different PR.

With this, users can run optitype on all sample types and get individual reports back from each. The priority of HLA typing results to be used for the actual epitope prediction is as follows:

``` ocaml
   (* HLA priority list
         - Manual HLAs
         - Seq2HLA results
         - OptiType on Normal DNA
         - OptiType on Tumor DNA
         - OptiType on Tumor RNA
    *)
```

**Comment by smondet** [comment]
LGTM, waiting on @ihodes  for merging :)

File: src/lib/pipeline.ml [hunk 3]
```diff
@@ -225,16 +228,21 @@ module Full (Bfx: Extended_edsl.Semantics) = struct
   let hla fqs =
```
Comment (ihodes) [inline src/lib/pipeline.ml]: might rename this "seq2hla" and make an equivalent one for Optitype. More importantly, need to add a Bfx.save for those optitypes, so they can be copied over into the final report.

**Comment by ihodes** [comment]
LGTM! One minor comment.

File: src/lib/pipeline.ml (hunk 3 above)
Comment (ihodes) [reply src/lib/pipeline.ml]: I take it back. It is saved. 

Then minor nit; could you rename the hla -> seq2hla_hla or something?

Comment (armish) [reply src/lib/pipeline.ml]: Yeah, good idea - done!

**Pull request merged**